langroid/agent/special/neo4j/neo4j_chat_agent.py

```diff
@@ -0,0 +1,293 @@
+        delete_query = """
+                MATCH (n)
+                DETACH DELETE n
+            """
+        if self.write_query(delete_query):
+            print("[green]Database is deleted!")
+        else:
+            print("[red]Database is not deleted!")
+
+    def make_query(self, msg: GenerateCypherQueries) -> str:
+        """ "
+        Handle a RunQueryTool message by executing a Cypher query and returning the
+        result.
+        Args:
+            msg (GenerateCypherQueries): The tool-message to handle.
+
+        Returns:
+            str: The result of executing the Cypherquery.
+        """
+        query = msg.cypherQuery
+
+        logger.info(f"Executing Cypher query: {query}")
+        return self.read_query(query)
+
+    # TODO: There are various ways to get the schema. The current one uses the func
+    # `read_query`, which requires post processing to identify whether the response upon
+    # the schema query is valid. Another way is to isolate this func from `read_query`.
+    # The current query works well. But we could use the queries here:
+    # https://github.com/neo4j/NaLLM/blob/1af09cd117ba0777d81075c597a5081583568f9f/api/
+    # src/driver/neo4j.py#L30
+    def get_schema(self, msg: GraphDatabaseSchema | None) -> str:
+        """
+        Retrieves the schema of a Neo4j graph database.
+
+        Args:
+            msg (GraphDatabaseSchema): An instance of GraphDatabaseSchema, typically
+            containing information or parameters needed for the database query.
+
+        Returns:
+            str: The visual representation of the database schema as a string, or a
+            message stating that the database schema is empty or not valid.
+
+        Raises:
+            This function does not explicitly raise exceptions but depends on the
+            behavior of 'self.read_query' method, which might raise exceptions related
+             to database connectivity or query execution.
+        """
+        schema = self.read_query("CALL db.schema.visualization()")
+        if not any(element in schema for element in not_valid_query_response):
+            return schema
+        else:
+            return "The database schema does not have any nodes or relationships."
+
+    def _init_message_tools(self) -> None:
+        """Initialize message tools used for chatting."""
+        message = self._format_message()
+        self.config.system_message = self.config.system_message.format(mode=message)
+        super().__init__(self.config)
+        self.enable_message(GenerateCypherQueries)
+        self.enable_message(GraphDatabaseSchema)
+
+    def _format_message(self) -> str:
+        if self.driver is None:
+            raise ValueError("Database driver None")
+
+        return (
+            SCHEMA_TOOLS_SYS_MSG.format(schema=self.get_schema(None))
+            if self.config.use_schema_tools
+            else DEFAULT_SYS_MSG
+        )
+
+    def agent_response(
```

Comment:
any reason this had to be overridden? Seems like this essentially copies the code of `ChatAgent.agent_response` and makes a special NEO4J error output? We should find a way to avoid this code duplication, maybe some of the task orch/signal mechanisms

Reply:
I tried here to mimic the implementation of `sql_chat_agent` https://github.com/langroid/langroid/blob/main/langroid/agent/special/sql/sql_chat_agent.py#L197